internal/scheduler/reports.go

```diff
@@ -6,441 +6,314 @@ import (
+	jobSchedulingContextByJobId := make(map[string]*JobSchedulingContext)
+	for queue, qctx := range sctx.QueueSchedulingContexts {
+		for jobId, jctx := range qctx.SuccessfulJobSchedulingContexts {
+			jobSchedulingContextByJobId[jobId] = jctx
+		}
+		for jobId, jctx := range qctx.UnsuccessfulJobSchedulingContexts {
+			jobSchedulingContextByJobId[jobId] = jctx
 		}
+		queueSchedulingContextByQueue[queue] = qctx
 	}
-	return &schedulerobjects.QueueReport{
-		Report: report.String(),
-	}, nil
+	return queueSchedulingContextByQueue, jobSchedulingContextByJobId
 }
 
-func (repo *SchedulingReportsRepository) GetJobReport(ctx context.Context, jobId *schedulerobjects.JobId) (*schedulerobjects.JobReport, error) {
-	jobUuid, err := uuidFromUlidString(jobId.Id)
-	if err != nil {
-		return nil, err
-	}
-	report, ok := repo.GetJobSchedulingReport(jobUuid)
-	if !ok {
-		return nil, &armadaerrors.ErrNotFound{
-			Type:    "JobSchedulingReport",
-			Value:   jobId.Id,
-			Message: "this job has not been considered for scheduling recently",
-		}
-	}
-	return &schedulerobjects.JobReport{
-		Report: report.String(),
+func (repo *SchedulingContextRepository) GetSchedulingReport(_ context.Context, _ *types.Empty) (*schedulerobjects.SchedulingReport, error) {
+	return &schedulerobjects.SchedulingReport{
+		Report: repo.getSchedulingReportString(),
 	}, nil
 }
 
-func (repo *SchedulingReportsRepository) AddSchedulingRoundReport(report *SchedulingRoundReport) {
-	for queue, queueSchedulingRoundReport := range report.QueueSchedulingRoundReports {
-		repo.AddMany(queue, maps.Values(queueSchedulingRoundReport.SuccessfulJobSchedulingReports))
-		repo.AddMany(queue, maps.Values(queueSchedulingRoundReport.UnsuccessfulJobSchedulingReports))
+func (repo *SchedulingContextRepository) getSchedulingReportString() string {
+	var sb strings.Builder
+	w := tabwriter.NewWriter(&sb, 1, 1, 1, ' ', 0)
+	for _, executorId := range repo.sortedExecutorIds {
+		fmt.Fprintf(w, "%s:\n", executorId)
+		sctx := repo.mostRecentSchedulingContextByExecutor[executorId]
+		if sctx != nil {
+			fmt.Fprint(w, indent.String("\t", "Most recent attempt:\n"))
+			fmt.Fprint(w, indent.String("\t\t", sctx.String()))
+		} else {
+			fmt.Fprint(w, indent.String("\t", "Most recent attempt: none\n"))
+		}
+		sctx = repo.mostRecentSuccessfulSchedulingContextByExecutor[executorId]
+		if sctx != nil {
+			fmt.Fprint(w, indent.String("\t", "Most recent successful attempt:\n"))
+			fmt.Fprint(w, indent.String("\t\t", sctx.String()))
+		} else {
+			fmt.Fprint(w, indent.String("\t", "Most recent successful attempt: none\n"))
+		}
 	}
+	w.Flush()
+	return sb.String()
 }
 
-func (repo *SchedulingReportsRepository) AddMany(queueName string, reports []*JobSchedulingReport) {
-	for _, report := range reports {
-		repo.Add(queueName, report)
-	}
+func (repo *SchedulingContextRepository) GetQueueReport(_ context.Context, queue *schedulerobjects.Queue) (*schedulerobjects.QueueReport, error) {
```

Comment:
Shouldn't there be a read lock around this (and also GetSchedulingReport)?

More generally, even if the logic here is correct, I think the mutable state protected by mutexes here is going to bight us sooner or later as this stuff is inherantly difficult to reason about and test.  I'd consider using another pattern here to try and make it safer. If it were me I'd probably try and create a state struct which held all the mutable state and then store that in an atomicReference. Writers would  deep copy the struct, update the copy and then store a new state.  Readers would just grab the struct.  It might also be possible to use iummutable maps int he struct which then might event remove the need to make a deep copy.  You'd still want a write mutex, but this would mean that any new read methods wouldn't need to worry about threading issues

Reply:
Agreed. I've changed it to use `atomic.Pointer` + deep-copy.